### DISCUSSION TITLE
Convert Cart To Order takes more than 25 seconds to perform

### DISCUSSION THREAD

**Discussion by balanh-development** [swellstores/community#505] [Help] [answered]
Currently every time I convert a cart into an order it takes a lot of time. Ex: CartId: 65d3aace87d07b0012604497
Previously, requests happened very quickly, our server is located in Singapore behind cloudflare. I think this is not a geographical issue, the API update order response is at an acceptable level of 1~2s. Webhooks when creating orders also respond very quickly. Any help is much appreciated.

**Comment by logeshswell** [comment]
1. Do you use real-time order webhooks?
If it takes a significant time to respond 200, then the cart conversion would also take the same time in addition.
2. Can you share your store URL to help you debug this issue?
3. Can you share a list of carts for which the issue happened?
It would be very helpful if you can share the API response time for the affected carts from the **Network** tab in the browser.

**Reply by ericingram** [reply]
You can also look Developer > Console > Logs to see the request time on Swell's end. I'm not sure if we are effectively logging the time for real-time webhooks yet but it does log related webhooks and soon app event hooks, which will more fully explain why a request takes a certain amount of time.

**Comment by balanh-development** [comment]
Hi @logeshswell , @ericingram , our team built gateway for swell, we monitored every request from our clients to swell server, here some detail

> PUT /carts/661a7973d321b30012078dee HTTP/1.1" 200 2486
> POST /orders HTTP/1.1" 200 2947 
> PUT /orders/661a7988c9f4750012f0def2 HTTP/1.1" 200 3084
> GET /accounts/64d7718f2a141f0011c2836e HTTP/1.1" 200 893
> GET /promotions?where%5Bactive%5D%5B$eq%5D=true&sort=date_created+asc HTTP/1.1" 200 3801

> "GET /orders/661a8512588bb400124688f4?include%5Bpayments%5D%5Burl%5D=payments&include%5Bpayments%5D%5Bparams%5D%5Border_id%5D=id&include%5Bpayments%5D%5Bdata%5D%5Blimit%5D=100 HTTP/1.1" 200 3144

Flow received carts -> register account -> update account-id to carts (1st call), convert carts to orders (2nd call), last update orders some addons info (3rd call)

Even, the get order api takes up to 3 seconds to execute

Api releated to carts, orders, promotions always takes more than 3, 4s to execute

**Reply by logeshswell** [reply]
> Even, the get order api takes up to 3 seconds to execute

`GET /orders/661a8512588bb400124688f4?expand=payments` should work comparatively better.

**Reply by logeshswell** [reply]
I will take a look at the rest of the requests that you shared and will update you if I get some updates.

**Reply by logeshswell** [reply]
@balanh-development From when did you notice this issue?

**Reply by logeshswell** [reply]
Do you experience the slow response time for order conversion always or when there's a higher traffic?

**Comment by logeshswell** [comment]
@balanh-development You mentioned that order conversion takes >25s in the GD title. But, all your shared logs show 3-4s, which is common as the requests might be throttled for non-enterprise clients, when there's a larger request traffic.

https://www.swell.is/help/faqs/pricing#what-does-fair-use-api-access-mean

**Reply by balanh-development** [reply]
Hi @logeshswell,
Our account: Standard - AppSumo Tier 4, it means there is no limit on response time ?

We are reviewing features internally in production, sure there are no high loads.
We have also just reproduce the make order order, logs detail below

GMT+0 (Numbers behind request: status, request_time, upstream_connect_time, upstream_header_time, upstream_response_time)

>[15/Apr/2024:13:26:15 +0000] "POST /orders HTTP/1.1" 500 (body: Maximum request time exceeded (45s))
>[15/Apr/2024:13:46:53 +0000] "POST /orders HTTP/1.1" 200 "30.706" "0.378" "30.707" "30.707"
>[15/Apr/2024:13:46:54 +0000] "PUT /orders/661d2faf3d55b20012fa386e HTTP/1.1" 200 "0.963" "0.381" "0.963" "0.963"
>[15/Apr/2024:13:46:55 +0000] "GET /orders/661d2faf3d55b20012fa386e?include%5Bpayments%5D%5Burl%5D=payments&include%5Bpayments%5D%5Bparams%5D%5Border_id%5D=id&include%5Bpayments%5D%5Bdata%5D%5Blimit%5D=100 HTTP/1.1" 200 "0.604" "0.379" "0.603" "0.603"


> [15/Apr/2024:13:50:46 +0000] "POST /orders HTTP/1.1" 200 "41.023" "0.371" "41.022" "41.023"
> [15/Apr/2024:13:50:47 +0000] "PUT /orders/661d308ea27705001266df9f HTTP/1.1" 200 "0.994" "0.371" "0.992" "0.993"

Note that 25s is the average time, and we have also optimized it so that we don't have to call swell too many steps. 
**But in the above case we just did, the cart has 4 products, and we cannot get response when converting cart.**

Your Swell's response, does it have any trace id from the header? We will log it and provide it to you, to make debugging easier.

**Reply by logeshswell** [reply]
> Our account: Standard - AppSumo Tier 4, it means there is no limit on response time ?

No, but you have higher limits than Starter and Basic plans.

**Reply by balanh-development** [reply]
Hi @logeshswell,
We cannot create a high workload, only a few testers are working on the system. You can also investigate right now, almost no trafic runs to swell services. Convert Cart API still takes more than 25s to make an order.

**Reply by logeshswell** [reply]
@balanh-development I could see a log throwing 500 error for order conversion in the last hour.

**Reply by logeshswell** [reply]
Can you help me with the API endpoint (or) curl request that you make?

Your API request doesn't seem to be the usual one where other merchants make an API request.

**Reply by logeshswell** [reply]
> We cannot create a high workload, only a few testers are working on the system

Sorry, I was wrong. It's not caused by higher traffic, as your requests are routed differently to our server.
I think something should be wrong at your API endpoint or other parameters.

**Reply by balanh-development** [reply]
Converting API only accepts field cartId, here is cartId: 661f96e76c03aa0012a9e722, we haven't convert it yet.

**Reply by logeshswell** [reply]
No, I was asking for API endpoint (or) curl request that you make.

**Reply by balanh-development** [reply]
We follow your documents: https://developers.swell.is/backend-api/carts/convert-carts-to-orders

curl https://api.swell.store/orders \
  -u store-id:secret-key \
  -d cart_id=661f96e76c03aa0012a9e722 \

**Reply by logeshswell** [reply]
I tried the above CURL request and it is returning error in real quick.
Please find the logs below:
```sh
$ curl -vs https://api.swell.store/orders -u store_id:secret_key -d cart_id=661f96e76c03aa0012a9e722
*   Trying 54.241.223.90:443...
* Connected to api.swell.store (54.241.223.90) port 443
* Server auth using Basic with user 'store_id'
> POST /orders HTTP/2
> Host: api.swell.store
> Authorization: Basic secret_key
> Content-Length: 32
> 
< HTTP/2 200 
< date: Fri, 19 Apr 2024 07:03:04 GMT
< content-type: application/json
< set-cookie: sticky-session=1713510185.244.26.741840|d6f48b8ace502b671f80228456abd007; Max-Age=3600; Path=/; Secure; HttpOnly
< strict-transport-security: max-age=15724800; includeSubDomains
< 
{"errors":{"cart":{"message":"Khoai Lang Nhật is not currently available","code":"INVALID"}}}
```

Do you always experience a slow API response for cart conversion?

**Reply by logeshswell** [reply]
At the same time, I could also see very few logs for which cart conversion requests timed out after 45s, while the rest of them responded very quickly.
Fetching products and stock values from the database caused this significant time delay. 
I informed my team about it and will keep you posted once the performance improvement is rolled out for your store.

**Reply by balanh-development** [reply]
Hi @logeshswell , your admin UI, does show "The product is out of stock"
![image](https://github.com/swellstores/community/assets/120219638/2fea11f3-fd37-4750-bc39-b95429c7cf37), I have removed the out of stock product from the cart, you can curl it again.

**Reply by logeshswell** [reply]
I am getting a different error in real quick.
```json
{"errors":{"cart":{"message":"Bánh Chưng Mặn is not currently available","code":"INVALID"}}}
```

**Reply by balanh-development** [reply]
please try again @logeshswell

**Comment by balanh-development** [comment]
"Do you always experience a slow API response for cart conversion?", yes currently we monitor converting API, it always takes more than 25 seconds to perform.

Here api example, one product in cart (Product is definitely in stock)

curl https://api.swell.store/orders
-u store-id:secret-key
-d cart_id=66223391868a6600122f70a9 \

**Reply by logeshswell** [reply]
Thanks for the details, it occurs very often.

**Reply by logeshswell** [reply]
I will investigate and update you the details of the fix.

**Reply by logeshswell** [reply]
Disabling your webhooks reduces the response time to 10s on an average. 
I also noticed that POST `/products:stocks` request consumes a significant time in your storefront and I am working on improving it.

**Reply by balanh-development** [reply]
We have the swell test version, no webhook integration, looks like it hasn't changed anything.

**Reply by logeshswell** [reply]
I was debugging in your live env. Can you check and confirm if your live env is working better?

**Reply by logeshswell** [reply]
In the meantime, I will take a look at your test env.

**Reply by balanh-development** [reply]
Hi @logeshswell, no changes yet, I just did the conversion cart, took 35 seconds

`/carts/662fcff219dfda00126070ca`

**Comment by logeshswell** [comment, marked as answer]
@balanh-development I raised a request internally to take a look at this issue. 
I doubt that it's a performance issue in creating the stock records during the cart conversion.

**Comment by logeshswell** [comment]
We made some fixes to improve the cart conversion process faster. Can you check and confirm if it's working faster now?
